spec/support/cmap.rb

```diff
@@ -0,0 +1,346 @@
+      attr_reader :target
+
+      # @return [ Array<Operation> ] thread_ops The operations to run on the thread.
+      attr_reader :thread_ops
+
+      # @return [ Integer | nil ] ms The number of milliseconds to sleep.
+      attr_reader :ms
+
+      # @return [ String | nil ] label The binding for a returned connection.
+      attr_reader :label
+
+      # @return [ String | nil ] The binding for the connection which should run the operation.
+      attr_reader :connection
+
+      # Create the new Operation.
+      #
+      # @example Create the new Operation.
+      #   Operation.new(operation)
+      #
+      # @param [ Hash ] operation The operation hash.
+      #
+      # @since 2.7.0
+      def initialize(operation)
+        @name = operation['name']
+        @thread = operation['thread']
+        @thread_ops = []
+        @target = operation['target']
+        @ms = operation['ms']
+        @label = operation['label']
+        @connection = operation['connection']
+      end
+
+      def run(state, main_thread = true)
+        case name
+        when 'start'
+          run_start_op(state)
+        when 'wait'
+          run_wait_op(state)
+        when 'waitFor'
+          run_wait_for_op(state)
+        when 'checkOut'
+          run_checkout_op(state)
+        when 'checkIn'
+          run_checkin_op(state)
+        when 'clear'
+          run_clear_op(state)
+        when 'close'
+          run_close_op(state)
+        else
+          raise "invalid operation: #{name}"
+        end
+
+        nil
+
+      # We hard-code the error messages because ours contain information like the address and the
+      # connection ID.
+      rescue Error::PoolClosed
+        raise unless main_thread
+
+        {
+          'type' => 'PoolClosedError',
+          'message' => 'Attempted to check out a connection from closed connection pool',
+        }
+      rescue Error::WaitQueueTimeout
+        raise unless main_thread
+
+        {
+          'type' => 'WaitQueueTimeoutError',
+          'message' => 'Timed out while checking out a connection from connection pool',
+        }
+      end
+
+      private
+
+      def run_start_op(state)
+        state[target] = Thread.start do
+          Thread.current[:name] = @target
+          thread_ops.each { |op| op.run(state, false) }
+        end
+
+        # Since we expect exceptions to occur in some cases, we disable the printing of error
+        # messages from the thread if the Ruby version supports it.
+        if state[target].respond_to?(:report_on_exception)
+          state[target].report_on_exception = false
+        end
+      end
+
+      def run_wait_op(_state)
+        sleep(ms / 1000.0)
+      end
+
+      def run_wait_for_op(state)
+        state[target].join
+      end
+
+      def run_checkout_op(state)
+        conn = state['pool'].checkout
+        state[label] = conn if label
+      end
+
+      def run_checkin_op(state)
+        until state[connection]; end
```

Comment:
This loop should have a sleep in it, otherwise it will starve the other threads which this thread is in fact waiting for.